sway/tree/arrange.c

```diff
@@ -0,0 +1,234 @@
+			root_container.sway_root->output_layout,
+			output->sway_output->wlr_output);
+	output->x = output_box->x;
+	output->y = output_box->y;
+	output->width = output_box->width;
+	output->height = output_box->height;
+	wlr_log(L_DEBUG, "Arranging output '%s' at %f,%f",
+			output->name, output->x, output->y);
+	for (int i = 0; i < output->children->length; ++i) {
+		struct sway_container *workspace = output->children->items[i];
+		arrange_workspace(workspace);
+	}
+	container_damage_whole(output);
+}
+
+void arrange_workspace(struct sway_container *workspace) {
+	if (config->reloading) {
+		return;
+	}
+	if (!sway_assert(workspace->type == C_WORKSPACE,
+			"called arrange_workspace() on non-workspace container")) {
+		return;
+	}
+	struct sway_container *output = workspace->parent;
+	struct wlr_box *area = &output->sway_output->usable_area;
+	wlr_log(L_DEBUG, "Usable area for ws: %dx%d@%d,%d",
+			area->width, area->height, area->x, area->y);
+	workspace->width = area->width;
+	workspace->height = area->height;
+	workspace->x = area->x;
+	workspace->y = area->y;
+	wlr_log(L_DEBUG, "Arranging workspace '%s' at %f, %f",
+			workspace->name, workspace->x, workspace->y);
+	arrange_children_of(workspace);
+	container_damage_whole(workspace);
+}
+
+static void apply_horiz_layout(struct sway_container *parent) {
+	size_t num_children = parent->children->length;
+	if (!num_children) {
+		return;
+	}
+	// Calculate total width of children
+	double scale = 0;
+	for (size_t i = 0; i < num_children; ++i) {
+		struct sway_container *child = parent->children->items[i];
+		if (child->width <= 0) {
+			if (num_children > 1) {
+				child->width = parent->width / (num_children - 1);
+			} else {
+				child->width = parent->width;
+			}
+		}
+		scale += child->width;
+	}
+	scale = parent->width / scale;
+
+	// Resize windows
+	wlr_log(L_DEBUG, "Arranging %p horizontally", parent);
+	double child_x = parent->x;
+	struct sway_container *child;
+	for (size_t i = 0; i < num_children; ++i) {
+		child = parent->children->items[i];
+		wlr_log(L_DEBUG,
+				"Calculating arrangement for %p:%d (will scale %f by %f)",
+				child, child->type, child->width, scale);
+		child->x = child_x;
+		child->y = parent->y;
+		child->width = floor(child->width * scale);
+		child->height = parent->height;
+		child_x += child->width;
+	}
+	// Make last child use remaining width of parent
+	child->width = parent->x + parent->width - child->x;
+}
+
+static void apply_vert_layout(struct sway_container *parent) {
+	size_t num_children = parent->children->length;
+	if (!num_children) {
+		return;
+	}
+	// Calculate total height of children
+	double scale = 0;
+	for (size_t i = 0; i < num_children; ++i) {
+		struct sway_container *child = parent->children->items[i];
+		if (child->height <= 0) {
+			if (num_children > 1) {
+				child->height = parent->height / (num_children - 1);
+			} else {
+				child->height = parent->height;
+			}
+		}
+		scale += child->height;
```

Comment:
It's confusing to use the `scale` variable to accumulate child heights. Maybe we could use another variable?